File: Data Structures & Algorithms/permutations/submission-0.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        permutations(nums, res, new ArrayList<>(), new boolean[nums.length]);
        return res;
    }

    private void permutations(int[] nums, List<List<Integer>> res, List<Integer> perm, boolean[] used) {        
        if (perm.size() == nums.length) {
            res.add(new ArrayList<>(perm));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (used[i]) continue;
            used[i] = true;
            perm.add(nums[i]);
            permutations(nums, res, perm, used);
            perm.remove(perm.size() - 1);
            used[i] = false;
        }
    }
}

File: Data Structures & Algorithms/subsets/submission-0.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> subset;
        vector<vector<int>> res;
        subsets(0, nums, res, subset);

        return res;
    }

    void subsets(int i, vector<int>& nums, vector<vector<int>>& res, vector<int>& subset) {
        if (i == nums.size()){
            res.push_back(subset);
            return;
        }
        

        subset.push_back(nums[i]);
        subsets(i+1, nums, res, subset);

        subset.pop_back();
        subsets(i+1, nums, res, subset);
    }


};
